File download/streaming

pascal260303 opened the pull request:
I found out how files can be downloaded/streamed and also how range requests work, so that the download url can be directly inserted in Foxglove to open the mcap file.

This uses the [django file storage api](https://docs.djangoproject.com/en/5.1/ref/files/storage/) for which a package [django-storages](https://django-storages.readthedocs.io/en/latest/) exists that brings support for S3 buckets and more.

An alternative for opening the mcap files in Foxglove would be generating [S3 pre-signed urls](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ShareObjectPreSignedURL.html) which is also possible and would give the user temporary direct access to the files in the S3 bucket.

This can also be used to stream the extracted video file from the backend to the frontend to render it there.

The url for downloading the files is http://localhost:8000/file/download/<file_path: path>.
When using local files, file_path is the relative path inside the backend/media folder.
When using S3 storage it is the path inside the S3 bucket.

resolves #181 

For authentication the sessionid must be added to the url as url param.

I added the fields `file_url` and `video_url` to the FileSerializer that contain the right url from where to download the file and already include the sessionid as url parameter.

I added a bit of documentation about files and the setup [here](https://github.com/brendel-group/mission_db/tree/feature/file-download/docs/files#readme).

I also added tests for the three kinds of file requests this supports.

I set the ruff version in requirements.txt and in the github action to the latest version to make sure they are the same version.

I had to set the django version in requirements.txt to a newer version, because the InMemoryStorage used for tests had a bug until this version. I could probably find a way around this bug with mock, since it only occurs in tests, if we want to support older django versions.

It's possible to create urls that directly open the foxlove app. An example would be : 
```url
foxglove://open?ds=remote-file&ds.url=http://localhost:8000/file/download/2024.10.14_picking_apples/test/bag_1728916594.8984177/bag_1728916594.8984177_0.mcap?sessionid=<sessionid>
```

## How to test
Make sure the example mission folders are in the `backend/media/` folder.\
Add all files and missions with the `sync` command of the CLI.\
Use the restapi to get the files of one mission.\
Open the `file_url` of one file in the browser to download the file or copy the url and open it in Foxglove as remote file.

It may be required to run this first:
```bash
python manage.py collectstatic
python manage.py makemigrations
python manage.py migrate
```

sh1negg reviewed (changes requested):
Hey @pascal260303,
both the implementation for using the `backend/media` directory, aswell as using S3 buckets work perfectly fine :) Generating the `file_url` also works when creating it locally and can be used e.g for establishing a connection with foxglove, however generating signed URLs doesn't work yet. 
![image](https://github.com/user-attachments/assets/55ad6549-ef8d-4187-90ac-671398b2fd1f)
Could you either please fix the issue and add the corresponding documentation for it or remove it from this PR and create a new issue handling this case?

pascal260303 commented:
I removed the comments in settings.py and added a bit of documentation about signed urls to [the documentation](https://github.com/brendel-group/mission_db/blob/feature/file-download/docs/files/README.md).\
I think whether we should use signed urls should be discussed in our next meeting.  So I wont open a new Issue before we discussed it.

sh1negg reviewed (dismissed):
Hey @pascal260303,
your new changes removed the possibility to generate any AWS links and now generates everything locally and it works as intended 👍 
I agree with you, that we should discuss the implementation of presigned urls in our next meeting, so for now, this PR fulfills all requirements from the linked issue #181 and can be merged :)

pascal260303 commented:
I also added a url and view for file streaming instead of just downloading because I was already trying #211 and it did download the file instead of playing the video. With this stream it works.\
It just changes two header fields of  the regular download response, so not much changed.

pascal260303 commented:
@sh1negg I had to resolve some merge conflicts can you approve again?
